pkg/core/klinedriver/klinebuilder.go

```diff
@@ -0,0 +1,230 @@
+		lastBucketStart := types.Time(nextBucket.StartTime.Time().Add(-accBucket.Interval.Duration()))
+		// setup the last kline and update the accumulating kline
+		var lastKLine types.KLine = *accBucket.KLine // make a copy
+		accBucket = kb.resetKLine(interval, nextBucket.StartTime)
+		if numShifts > 1 {
+			lastKLine.Open = lastKLine.Close
+			lastKLine.High = lastKLine.Close
+			lastKLine.Low = lastKLine.Close
+			lastKLine.Volume = fixedpoint.Zero
+			lastKLine.QuoteVolume = fixedpoint.Zero
+			lastKLine.TakerBuyBaseAssetVolume = fixedpoint.Zero
+			lastKLine.TakerBuyQuoteAssetVolume = fixedpoint.Zero
+			lastKLine.NumberOfTrades = 0
+			lastKLine.StartTime = lastBucketStart
+			lastKLine.EndTime = nextBucket.StartTime
+		}
+		lastKLine.Closed = true
+		lastBucket := &Bucket{
+			StartTime: lastBucketStart,
+			EndTime:   nextBucket.StartTime,
+			Interval:  interval,
+			KLine:     &lastKLine,
+		}
+		kb.lastBucketMap[interval] = lastBucket
+		accumulateKLineInPlace(accBucket.KLine, trade)
+	}
+}
+
+// Update updates the KLineBuilder with the given update time.
+// The semantics of the Update is to notify the current timestamp to the KLineBuilder
+func (kb *KLineBuilder) Update(updateTime types.Time) (kLinesMap map[types.Interval]types.KLine) {
+	kLinesMap = make(map[types.Interval]types.KLine)
+	for interval, accBucket := range kb.accBucketMap {
+		lastBucket, hasLast := kb.lastBucketMap[interval]
+		if accBucket.KLine.NumberOfTrades == 0 && !hasLast {
+			// no last kline and no trades happened in this interval, no kline to return
+			log.Warnf("kline not available for interval %s(%s) yet", kb.symbol, interval)
+			continue
+		}
+		// either there were trades or we have a last kline
+		var kline types.KLine // the kline to be returned
+		kline.Set(accBucket.KLine)
+		if accBucket.KLine.NumberOfTrades == 0 {
+			// no trade happened up to the updateTime
+			// use the last kline to fill the gap
+			kline.Open = lastBucket.KLine.Close
+			kline.High = lastBucket.KLine.Close
+			kline.Low = lastBucket.KLine.Close
+			kline.Close = lastBucket.KLine.Close
+			kline.LastTradeID = lastBucket.KLine.LastTradeID
+			kline.Volume = fixedpoint.Zero
+			kline.QuoteVolume = fixedpoint.Zero
+			kline.TakerBuyBaseAssetVolume = fixedpoint.Zero
+			kline.TakerBuyQuoteAssetVolume = fixedpoint.Zero
+		}
+		// check if we should close the kline
+		checkClosedInPlace(&kline, updateTime, accBucket)
+		if kline.Closed {
+			nextBucket, _ := findNextBucket(updateTime, accBucket)
+			resetBucket := kb.resetKLine(interval, nextBucket.StartTime)
+			resetBucket.KLine.Open = kline.Close
+			resetBucket.KLine.High = kline.Close
+			resetBucket.KLine.Low = kline.Close
+			resetBucket.KLine.Close = kline.Close
+			resetBucket.KLine.LastTradeID = kline.LastTradeID
+		}
+		kLinesMap[interval] = kline
+		kb.lastBucketMap[interval] = accBucket
+	}
+	return
+}
+
+func (kb *KLineBuilder) resetKLine(interval types.Interval, resetTime types.Time) *Bucket {
```

Comment:
can call it `resetInterval` since the input is "interval"

Reply:
Done.